File: tests/java/tests-java-junit5-api_request-playwright/src/test/java/api/PlaywrightHttp.java
package api;

import allure.AllureApiRequest;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.microsoft.playwright.APIRequest;
import com.microsoft.playwright.APIRequestContext;
import com.microsoft.playwright.APIResponse;
import com.microsoft.playwright.Playwright;
import com.microsoft.playwright.options.RequestOptions;
import config.ConfigReader;
import config.TestConfig;

import java.util.Locale;

/**
 * Shared Playwright {@code APIRequestContext} against {@link ConfigReader#resolveApiBaseUrl()}.
 */
public final class PlaywrightHttp {

    public static final String WRONG_CREDENTIALS_MESSAGE = "Wrong login or password";

    private static final Object GATE = new Object();

    private static Playwright playwright;
    private static APIRequestContext api;
    private static String origin = "";
    private static TestConfig config;

    private PlaywrightHttp() {
    }

    public static void setup(TestConfig testConfig) {
        synchronized (GATE) {
            if (api != null) {
                return;
            }
            config = testConfig;
            origin = ConfigReader.resolveApiBaseUrl().replaceAll("/$", "");
            playwright = Playwright.create();
            api = playwright.request().newContext(new APIRequest.NewContextOptions()
                    .setBaseURL(origin)
                    .setTimeout(10_000)
                    .setIgnoreHTTPSErrors(true));
            Runtime.getRuntime().addShutdownHook(new Thread(PlaywrightHttp::close));
        }
    }

    public static void ensure() {
        setup(ConfigReader.testConfig);
    }

    public static HttpResult request(String method, String path, Object json, String raw, String token) {
        ensure();
        RequestOptions options = RequestOptions.create();
        if (token != null) {
            options.setHeader("Authorization", "Bearer " + token);
        }
        if (json != null) {
            options.setHeader("Content-Type", "application/json");
            try {
                options.setData(HttpResult.MAPPER.writeValueAsString(json));
            } catch (JsonProcessingException e) {
                throw new IllegalArgumentException(e);
            }
        } else if (raw != null) {
            options.setHeader("Content-Type", "application/json");
            options.setData(raw);
        }
        String verb = method.toUpperCase(Locale.ROOT);
        String urlPath = pathOf(path);
        APIResponse response = switch (verb) {
            case "GET" -> api.get(urlPath, options);
            case "POST" -> api.post(urlPath, options);
            case "PUT" -> api.put(urlPath, options);
            case "DELETE" -> api.delete(urlPath, options);
            default -> throw new IllegalArgumentException(verb);
        };
        try {
            int status = response.status();
            String body = response.text();
            AllureApiRequest.attach(config, verb, origin + urlPath, status, body);
            return new HttpResult(status, body);
        } finally {
            response.dispose();
        }
    }

    public static HttpResult request(String method, String path) {
        return request(method, path, null, null, null);
    }

    public static HttpResult request(String method, String path, Object json) {
        return request(method, path, json, null, null);
    }

    public static HttpResult request(String method, String path, String token) {
        return request(method, path, null, null, token);
    }

    static void close() {
        synchronized (GATE) {
            if (api != null) {
                api.dispose();
                api = null;
            }
            if (playwright != null) {
                playwright.close();
                playwright = null;
            }
        }
    }

    private static String pathOf(String path) {
        return path.startsWith("/") ? path : "/" + path;
    }
}

File: tests/kotlin/tests-kotlin-junit5-api_request-playwright/src/test/kotlin/api/PlaywrightHttp.kt
package api

import com.microsoft.playwright.APIRequest
import com.microsoft.playwright.APIRequestContext
import com.microsoft.playwright.Playwright
import com.microsoft.playwright.options.RequestOptions
import config.ConfigReader
import config.TestConfig
import io.qameta.allure.Allure

/**
 * Shared Playwright `APIRequestContext` against [ConfigReader.resolveApiBaseUrl].
 */
object PlaywrightHttp {

    const val WRONG_CREDENTIALS_MESSAGE: String = "Wrong login or password"

    @Volatile
    private var playwright: Playwright? = null

    @Volatile
    private var api: APIRequestContext? = null

    @Volatile
    private var origin: String = ""

    @Volatile
    private var config: TestConfig? = null

    @Synchronized
    fun setup(testConfig: TestConfig) {
        if (api != null) {
            return
        }
        config = testConfig
        origin = ConfigReader.resolveApiBaseUrl().trimEnd('/')
        val pw = Playwright.create()
        playwright = pw
        api = pw.request().newContext(
            APIRequest.NewContextOptions()
                .setBaseURL(origin)
                .setTimeout(10_000.0)
                .setIgnoreHTTPSErrors(true),
        )
        Runtime.getRuntime().addShutdownHook(Thread { close() })
    }

    fun request(
        method: String,
        path: String,
        json: Any? = null,
        raw: String? = null,
        token: String? = null,
    ): HttpResult {
        val http = checkNotNull(api) { "PlaywrightHttp.setup() first" }
        val options = RequestOptions.create()
        if (token != null) {
            options.setHeader("Authorization", "Bearer $token")
        }
        when {
            json != null -> {
                options.setHeader("Content-Type", "application/json")
                options.setData(HttpResult.MAPPER.writeValueAsString(json))
            }
            raw != null -> {
                options.setHeader("Content-Type", "application/json")
                options.setData(raw)
            }
        }
        val verb = method.uppercase()
        val urlPath = pathOf(path)
        val response = when (verb) {
            "GET" -> http.get(urlPath, options)
            "POST" -> http.post(urlPath, options)
            "PUT" -> http.put(urlPath, options)
            "DELETE" -> http.delete(urlPath, options)
            else -> error(verb)
        }
        try {
            val result = HttpResult(response.status(), response.text())
            attach(verb, origin + urlPath, result.status, result.body)
            return result
        } finally {
            response.dispose()
        }
    }

    private fun attach(method: String, url: String, status: Int, body: String) {
        val cfg = config ?: return
        if (cfg.allureReportMode() == "none" || !cfg.enableAllureRestAssuredListener()) {
            return
        }
        try {
            Allure.addAttachment("Request", "text/plain", "$method $url")
            Allure.addAttachment("Response", "text/plain", "$status\n$body")
        } catch (_: RuntimeException) {
            // Allure context is optional; never mask the HTTP call.
        }
    }

    @Synchronized
    private fun close() {
        api?.dispose()
        api = null
        playwright?.close()
        playwright = null
    }

    private fun pathOf(path: String): String = if (path.startsWith("/")) path else "/$path"
}
